File: src/main/java/com/example/ajaxonboardingproject/controller/UserController.java
package com.example.ajaxonboardingproject.controller;

import com.example.ajaxonboardingproject.dto.response.UserResponseDto;
import com.example.ajaxonboardingproject.model.User;
import com.example.ajaxonboardingproject.service.UserService;
import com.example.ajaxonboardingproject.service.mapper.ResponseDtoMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

@RestController
@RequiredArgsConstructor
@RequestMapping("/users")
public class UserController {
    private final UserService userService;
    private final ResponseDtoMapper<UserResponseDto, User> userResponseDtoMapper;

    @GetMapping("/by-email")
    public UserResponseDto findByEmail(@RequestParam String email) {
        User user = userService.findByEmail(email).orElseThrow(
                () -> new RuntimeException("User with email " + email + " not found"));
        return userResponseDtoMapper.mapToDto(user);
    }
}

File: src/main/java/com/example/ajaxonboardingproject/controller/UserController.kt
package com.example.ajaxonboardingproject.controller

import com.example.ajaxonboardingproject.dto.response.UserResponseDto
import com.example.ajaxonboardingproject.model.User
import com.example.ajaxonboardingproject.service.UserService
import com.example.ajaxonboardingproject.service.mapper.ResponseDtoMapper
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/users")
data class UserController(
    private val userService: UserService,
    private val userResponseDtoMapper: ResponseDtoMapper<UserResponseDto, User>
) {
    @GetMapping("/by-email")
    fun finByEmail(@RequestParam email: String): UserResponseDto {
        return userService.findByEmail(email)
            .map { userResponseDtoMapper.mapToDto(it) }
            .block()!!
    }
}
